src/RequestSender.ts

```diff
@@ -0,0 +1,547 @@
+    //   no Transfer-Encoding is sent and the request method defines a meaning
+    //   for an enclosed payload body.  For example, a Content-Length header
+    //   field is normally sent in a POST request even when the value is 0
+    //   (indicating an empty payload body).  A user agent SHOULD NOT send a
+    //   Content-Length header field when the request message does not contain
+    //   a payload body and the method semantics do not anticipate such a
+    //   body.
+    //
+    // These method types are expected to have bodies and so we should always
+    // include a Content-Length.
+    const methodHasPayload =
+      method == 'POST' || method == 'PUT' || method == 'PATCH';
+
+    // If a content length was specified, we always include it regardless of
+    // whether the method semantics anticipate such a body. This keeps us
+    // consistent with historical behavior. We do however want to warn on this
+    // and fix these cases as they are semantically incorrect.
+    if (methodHasPayload || contentLength) {
+      if (!methodHasPayload) {
+        utils.emitWarning(
+          `${method} method had non-zero contentLength but no payload is expected for this verb`
+        );
+      }
+      defaultHeaders['Content-Length'] = contentLength;
+    }
+
+    return Object.assign(
+      utils.removeNullish(defaultHeaders),
+      // If the user supplied, say 'idempotency-key', override instead of appending by ensuring caps are the same.
+      utils.normalizeHeaders(userSuppliedHeaders)
+    );
+  }
+
+  _getUserAgentString(): string {
+    const packageVersion = this._stripe.getConstant('PACKAGE_VERSION');
+    const appInfo = this._stripe._appInfo
+      ? this._stripe.getAppInfoAsString()
+      : '';
+
+    return `Stripe/v1 NodeBindings/${packageVersion} ${appInfo}`.trim();
+  }
+
+  _getTelemetryHeader(): string | undefined {
+    if (
+      this._stripe.getTelemetryEnabled() &&
+      this._stripe._prevRequestMetrics.length > 0
+    ) {
+      const metrics = this._stripe._prevRequestMetrics.shift();
+      return JSON.stringify({
+        last_request_metrics: metrics,
+      });
+    }
+  }
+
+  _recordRequestMetrics(requestId: string, requestDurationMs: number): void {
+    if (this._stripe.getTelemetryEnabled() && requestId) {
+      if (
+        this._stripe._prevRequestMetrics.length > this._maxBufferedRequestMetric
+      ) {
+        utils.emitWarning(
+          'Request metrics buffer is full, dropping telemetry message.'
+        );
+      } else {
+        this._stripe._prevRequestMetrics.push({
+          request_id: requestId,
+          request_duration_ms: requestDurationMs,
+        });
+      }
+    }
+  }
+
+  _request(
+    method: string,
+    host: string | null,
+    path: string,
+    data: RequestData,
+    auth: string | null,
+    options: RequestOptions = {},
+    callback: RequestCallback,
+    requestDataProcessor: RequestDataProcessor | null = null
```

Comment:
Is there a reason why we don't use `requestDataProcessor?: RequestDataProcessor` instead here?

Reply:
Nevermind, I see we don't want it to be `'undefined'`